examples/multisite_neuroimg_adapt/main.py

```diff
@@ -0,0 +1,129 @@
+"""
+Autism Detection: Domain Adaptation for Multi-Site Neuroimaging Data Analysis
+
+Reference:
+[1] Cameron Craddock, Yassine Benhajali, Carlton Chu, Francois Chouinard, Alan Evans, András Jakab, Budhachandra Singh Khundrakpam, John David Lewis, Qingyang Li, Michael Milham, Chaogan Yan, Pierre Bellec (2013). The Neuro Bureau Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives. In Neuroinformatics 2013, Stockholm, Sweden.
+
+[2] Abraham, A., Pedregosa, F., Eickenberg, M., Gervais, P., Mueller, A., Kossaifi, J., ... & Varoquaux, G. (2014). Machine learning for neuroimaging with scikit-learn. Frontiers in neuroinformatics, 14.
+
+[3] Zhou, S., Li, W., Cox, C.R., & Lu, H. (2020). Side Information Dependence as a Regularizer for Analyzing Human Brain Conditions across Cognitive Experiments. in AAAI 2020, New York, USA. https://ojs.aaai.org//index.php/AAAI/article/view/6179
+
+[4] Zhou, S. (2022). Interpretable Domain-Aware Learning for Neuroimage Classification (Doctoral dissertation, University of Sheffield). https://etheses.whiterose.ac.uk/31044/1/PhD_thesis_ShuoZhou_170272834.pdf
+"""
+import os
+import warnings
+
+import numpy as np
+import pandas as pd
+from config import get_cfg_defaults
+from nilearn.connectome import ConnectivityMeasure
+from nilearn.datasets import fetch_abide_pcp
+from sklearn.linear_model import RidgeClassifier
+from sklearn.metrics import accuracy_score
+from sklearn.preprocessing import OneHotEncoder
+
+from kale.interpret import visualize
+from kale.pipeline.multi_domain_adapter import CoIRLS
+from kale.utils.download import download_file_by_url
+
+
+def cross_validation(x, y, covariates, estimator, domain_adaptation=False):
+    results = {"Target": [], "Num_samples": [], "Accuracy": []}
+    unique_covariates = np.unique(covariates)
+    n_covariates = len(unique_covariates)
+    enc = OneHotEncoder(handle_unknown="ignore")
+    covariate_mat = enc.fit_transform(covariates.reshape(-1, 1)).toarray()
+
+    for tgt in unique_covariates:
+        idx_tgt = np.where(covariates == tgt)
+        idx_src = np.where(covariates != tgt)
+        x_tgt = x[idx_tgt]
+        x_src = x[idx_src]
+        y_tgt = y[idx_tgt]
+        y_src = y[idx_src]
+
+        if domain_adaptation:
+            estimator.fit(
+                np.concatenate((x_src, x_tgt)), y_src, np.concatenate((covariate_mat[idx_src], covariate_mat[idx_tgt]))
+            )
+        else:
+            estimator.fit(x_src, y_src)
+        y_pred = estimator.predict(x_tgt)
+        results["Accuracy"].append(accuracy_score(y_tgt, y_pred))
+        results["Target"].append(tgt)
+        results["Num_samples"].append(x_tgt.shape[0])
+
+    mean_acc = sum([results["Num_samples"][i] * results["Accuracy"][i] for i in range(n_covariates)])
+    mean_acc /= x.shape[0]
+    results["Target"].append("Average")
+    results["Num_samples"].append(x.shape[0])
+    results["Accuracy"].append(mean_acc)
+
+    return pd.DataFrame(results)
```

Comment:
As discussed, this should go into `kale.evaluate`

Reply:
A module `cross_validation` can be created under `kale.evaluate`. This function can be called `leave_one_group_out_cross_validate`.

The source code of these two APIs in scikit-learn can be helpful:

- [LeaveOneGroupOut](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.LeaveOneGroupOut.html#sklearn.model_selection.LeaveOneGroupOut)
- [cross_validate](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate)